### ISSUE TITLE
Games with MLAA implementations can't have their resolutions properly upscaled

### ISSUE THREAD

**Issue by B-Knight49** [RPCS3/rpcs3#6348]
**System Information**
rpcs3-v0.0.6-8464-0a5b6ad9

i9-9900K @ 4.70 GHz
16GB RAM
GTX 980Ti
Windows 10 Build 1803

I'll attach a log of LBP2 if required (Lemme know!). 
It may be several hundred MB's though.

**Issue**
Games such as God Of War 3 and LittleBigPlanet 2 (only two I can test currently) can not have their resolutions upscaled without artefacting, blurriness or basically no improvement to resolution. It seems like only text and certain elements are upscaled making it only rarely sharper and clearer. Not sure of its reliability but [this list](https://forum.beyond3d.com/posts/1114422/) shows what games do/don't have MLAA. Anyone able to test any of the others?

In the case of LBP2, performance is hugely negatively affected by having MLAA enabled. Disabling it by using a leaked debug build of the game results in a consistent 30FPS, better image quality and, if VSYNC is disabled, a consistent 60FPS. This is obviously because MLAA is processed on the CPU. In-game info shows CPU timings dropping 50% with MLAA disabled [See screenshots]. Ignore the game ID of LBP2. I made two installs and randomly gave it the ID BCES01080 in order to separate my real and debug installations.

**God of War 3 (500% // 6400x3600)** 500% to show effects easier

![image](https://user-images.githubusercontent.com/26165117/63093934-d9989f80-bf5e-11e9-8156-4ec20b8d5820.png)

**Sharper text and smoother edges on pause screen**. Minor artefacting with mist/fog in background
![image](https://user-images.githubusercontent.com/26165117/63093989-fdf47c00-bf5e-11e9-9b05-b13d92ce5af4.png)

**LittleBigPlanet 2 [Debug] (200% // 2560x1440)**
**MLAA enabled**
![image](https://user-images.githubusercontent.com/26165117/63094330-f5e90c00-bf5f-11e9-9f33-025f6b793246.png)

**MLAA disabled**. Infinitely better image quality
![image](https://user-images.githubusercontent.com/26165117/63094361-0dc09000-bf60-11e9-9efd-6ffb393b66f6.png)

**Proof it's MLAA**
![image](https://user-images.githubusercontent.com/26165117/63094408-36e12080-bf60-11e9-9828-1de513d44f6b.png)

**CPU/GPU Timings**
![image](https://user-images.githubusercontent.com/26165117/63094877-8411c200-bf61-11e9-8f0d-8d5d09be2a20.png)

I mentioned this on the RPCS3 discord but thought I'd post it here too. At the time I was told that it could never be fixed and is reliant on game-specific patches. However, seeing its significant effects both on performance and image quality, I can't help but feel as though this needs its own issue and discussion to hopefully find a better solution. If fixed, games could have their resolutions upgraded to PC-tier quality and performance could improve by up to 25% or more (if LBP2 is anything to go off of).

**Comment by B-Knight49** [comment]
Portal 2 is an interesting one. The image becomes sharper / less blurry but anti-aliasing seems to have completely gone out of the window. There's also a weird graphical effect where there are 'waves' on some textures - this is most easily seen on the back wall. Seems like this is the extent of the issue here. Interesting that edges become extremely jagged at such a high resolution though, maybe related to the MLAA reading the (upscaled) frame incorrectly?

Both images use an internal resolution of 1080p. Results were identical at 720p.

**Portal 2**
![image](https://user-images.githubusercontent.com/26165117/63108358-8fbfb180-bf7e-11e9-8464-3c497fb216bc.png)

**Portal 2 (500% // 6400x3600)** 500% to show effects easier
![image](https://user-images.githubusercontent.com/26165117/63108408-aebe4380-bf7e-11e9-95c0-de9f8b486ba3.png)

**1080p (100%) Zoomed**
![image](https://user-images.githubusercontent.com/26165117/63108838-bcc09400-bf7f-11e9-9f47-bfa5c2740d99.png)

**3600p (500%) Zoomed**
![image](https://user-images.githubusercontent.com/26165117/63108702-6ce1cd00-bf7f-11e9-8f80-04534c331801.png)

**Comment by kd-11** [comment]
All these games will use different techniques for AA. Usually compute-heavy passes are offloaded to the SPUs (you can see how the work shifts away from the GPU when it is enabled) and we have to supply a downsampled image to fit the memory allocation. Outside of patching the binaries to remove the passes, I'm not seeing another solution for things like this.

**Comment by B-Knight49** [comment]
Yeah, that was the general consensus from the discord too. As I say though, since it has such a significant impact on both graphical fidelity and performance, I felt I just had to make an issue. Hopefully someone can be creative or, in the future, think of a real solution. 

I'm in no position to offer any sort of real advice but I am curious, so forgive my ignorance: could RPCS3 not even aid in or automate the creation of a patch? Maybe through memory editing similar to cheat engine? Perhaps even offer a config option that might disable AA instructions/calls on a per-game basis? 

I'm not sure. I made this to ignite a discussion and hopefully lead to something is all.

**Comment by RainbowCookie32** [comment]
You may want to check out #6038 for that

**Comment by B-Knight49** [comment]
Yeah, that's what I was thinking of. The only issue is it's not permanent I guess. I suppose future additions could allow for things like cheat tables and pointers etc but that'd be a decent amount of work, no?

**Comment by isJuhn** [comment]
RPCS3's patch system is enough, but you'd need someone to write the patch too. By the help of that lbp2 debug build it might not be so hard but in other games I can imagine this being quite the headache to get right.

**Comment by B-Knight49** [comment]
Yeah, I've only an extremely basic understanding of assembler and memory editing but I can't imagine it'd be easy to find the value that controls things like this. With LBP2 it might not be so bad since there's clearly a toggle for it purely as a result of the debug build. If the internals of other games contain no such toggle, patches and memory editing will have to read/edit hundreds if not thousands of things, right?

**Comment by kd-11** [comment]
There is no such thing as AA instructions so this is the only way. RSX emulation is not really HLE'd by patching because the eboot has bits of the driver inlined into the code. In short, modifying operation of the EBOOT is the only way to get around this. Now, there has always been the option of doing shader replacement, but in these games that read back the values into CPU and then process the image there, this would have no effect.

**Comment by B-Knight49** [comment]
That's unfortunate. I understand about other AA techniques but, just to clarify, isn't MLAA done completely by the CPU though? Not that I imagine it really changes much.

I'd like to request to the mods to keep this issue open for now though. Again, this is pretty significant so any other suggestions could be really useful.

**Comment by kd-11** [comment]
Yea, doesn't change much. From the POV of a processor (GPU or CPU) these effects are just basic arithmetic instruction sequences; we can't know its MLAA just like your physical CPU cannot know what you're actually doing on your desktop or in a game, etc.

**Comment by B-Knight49** [comment]
It's just occurred to me that supporting, encouraging or sharing modified EBOOT's isn't allowed since it's a copyrighted file - which further complicates matters. That means we've gotta rely on RPCS3 to do things in-house, right?

That sorta brings me back to something I briefly mentioned earlier: could the emulator make a patched binary by itself? AFAIK, modifying an EBOOT is done through decrypting it, using a hex editor and then resigning it. Here's something I've just brainstormed that, if it could work, would be a lot of effort to implement but could fix this issue and open up a whole new realm of possibility for modifications - it'd do more than just fix this issue. Basically, we have RPCS3 automate the creation of a patched binary so we can prevent copyright issues since it relies on the user having the EBOOT in the first place. 

0. The game's original EBOOT doesn't get modified. We instead copy it, modify it and then (if possible) provide a way to use the modified EBOOT in-place of the original by using a pointer or link? If you're familiar with it, this is akin to what Nexus Mod Manager does. I've never seen an EBOOT greater than a few MB so copying the file shouldn't be too bad either. If the linking/pointer doesn't work, we could always back-up the original somewhere safe and then place the modified one in the USRDIR folder, removing the link or moving the OG file back whenever the user wants. For extra "user-friendly" points, a checkbox could be added to the game's custom configuration saying something like "Use custom offsets" and, if necessary, providing a pop-up warning of side effects.

1. Every game should have a custom file (similar to patch.yml) that is provided with offsets / bytes that the community can provide through Reddit/Forums/etc. Of course, these offsets/bytes will vary per-game and even per-version and this would have to be noted in the post title - I've seen this done before with things like iPhone exploits when jailbreaking (Home_Depot). Of course this could lead to issues with the game booting and data corruption if the wrong data is inputted but this is just as much of a risk as with patch.yml I assume, right?

2. RPCS3 recognises the file (let's call it offsets.yml) and decrypts the EBOOT after it has been copied (in Step 0).

3. Using the data provided in offsets.yml, RPCS3 (somehow) hex edits the decrypted EBOOT making whatever change the user desires. This can be anything from cheats, mod menus to game-specific fixes like this topic.

4. RPCS3 resigns the modified EBOOT and uses that newly modified one to start the game. If everything goes well, we now have a modified game that uses a process which is (mostly) user-friendly and legal to support. It is, however, gonna be a pain to code and implement - sadly I've not got the experience to even offer help here.

What are the odds of something like this coming to fruition? Is it theoretically possible? If so, it'd undoubtedly have to be a WIP, additional feature like #6038 that is done over time and absolutely not a priority. Anyone got any thoughts?

**Comment by isJuhn** [comment]
Isn't this just patch.yml but a bit more user friendly? patch.yml is a file with first a hash which identifies the executable, and then it's just a series of hex edits: [type, offset, value]. rpcs3 decrypts and loads the executable into memory, then it applies all the hex edits from the patch file and then it launches the emulation. These patches are distributed on the rpcs3 github [here](https://github.com/RPCS3/rpcs3/wiki/Game-Patches) and just through users on any medium.

There are of course improvements to be made on the user friendly side, like adding a UI for selecting which patches you want to enable etc. But right now it's good enough for something like this.

**Comment by B-Knight49** [comment]
Correct me if I'm wrong but that's just memory editing and doesn't work for things like this. As kd-11 said, you'd need to manually edit the EBOOT files for changes like this. Of course, if you think you could get it to work using the current system, I'd love to know how!

**Comment by isJuhn** [comment]
it's for editing the executable without changing the file itself and it will work perfectly for stuff like this

**Comment by B-Knight49** [comment]
Well, I'm intrigued now. You're certain this could work for all games, though? kd-11 seems to be pretty set on the fact this isn't gonna be solved without EBOOT edits.

**Comment by Whatcookie** [comment]
EBOOT.BIN is the executable. You can edit EBOOT.BIN with patch.yml

**Comment by B-Knight49** [comment]
Hey Whatcookie, you're the one who responded to me on Discord originally. Hope you didn't interpret this post as me ignoring what you said, I just thought I'd see if there could be a greater discussion in a more centralised place. 

But maybe I misinterpreted this then and realise now my suggestion is literally just patch.yml lol. I assumed that "EBOOT edits" meant manual hex editing using something like HxD and I assumed patch.yml was memory editing... or something. Again, excuse my ignorance and lack of knowledge. 

Nonetheless, sucks that this is still patch.yml dependent. But knowing it's simpler than I originally thought is comforting, even if it's still slightly annoying though.

**Comment by kd-11** [comment]
If this is about a nice way of altering game logic then that is misguided and impossible. There is simply no way of altering the behavior of a game universally without changing either the bytes in the eboot (via hex edit or patch.yml which does the same thing but easier as it is a drag-drop solution) or changing the game's other files/assets via a mod. Either way, someone will have to painstakingly go through the bytestream and identify what needs changing. It will always be an intrusive process since you're literally going against what the developer of the program asked the program to do.

**Comment by lextra2** [comment]
Didn't know FF13 uses MLAA on PS3. I thought it used MSAA. It for sure uses MSAA on PC since it only allows it to be set to a minimum of 2x in the options. And forcing it off via gpu driver menu shows you why: horrible eyelashes and brows since it relies on it for proper alpha transparency.

**Comment by kd-11** [comment]
This ticket is not about MLAA per se, its about any game that uses CPU or 2D acceleration for post processing. I'll change the title to reflect this.

**Comment by lextra2** [comment]
> This ticket is not about MLAA per se, its about any game that uses CPU or 2D acceleration for post processing. I'll change the title to reflect this.

alright. fair enough.

**Comment by B-Knight49** [comment]
@lextra2 In your new ticket you mention that disabling AA fixed the artifacting you were getting. I'm interested in why this doesn't work for the games I mentioned in this post though? Even with anti-aliasing off, the AA isn't _actually disabled_ and I still get the same upscaling problems.

**Comment by B-Knight49** [comment]
Feel compelled to mention #7211 here since it addresses the performance problems MLAA brought. Thanks to Whatcookie. Hopefully this'll lead to resolution upscaling in MLAA games in the future too!

**Comment by Darkhost1999** [comment]
Any update on this issue? Last comment appears to be 2 years old. Which games are still effected? Any changes?

**Comment by AniLeo** [comment]
You can now use FSR as a way to work around that

The other way is disabling MLAA through a game patch, which most if not all games that use it should already have

**Comment by MSuih** [comment]
There was some improvements that improved upscaling compatibility, they were showcased in [this video](https://www.youtube.com/watch?v=_aEVRkxR08k).

**Comment by nsoeth** [comment]
> You can now use FSR as a way to work around that
> 
> The other way is disabling MLAA through a game patch, which most if not all games that use it should already have

Hmm, the problem exists i.e. in NHL 14 and NHL Legacy Edition. NHL 12 works fine. Anybody could help with this? I don't have a lot (none) experience with byte-editing.

**Comment by UnionExxis** [comment]
Sorry to necro this thread, but I think Captain America: Super Soldier might be one of these games that doesn't scale well. I am using 200% resolution scale to attempt to run the game at 1440p, there's some weird artifacts happening. Text in the UI seems upscaled as it is nice and sharp. Example here, look at edges of the helmet along his cheekbones and the fence in the background:

![FPS_ 53 44 _ Vulkan _ 0 0 21-13380 _ Captain America™_ Super Soldier  BLES01167  3_28_2022 6_31_09 PM](https://user-images.githubusercontent.com/73078011/160506817-ade0ff4d-0970-4de3-a280-3bdc7a12a496.png)

FSR certainly helps as an alterative, but even at 50% the image seems soft and text starts to take a hit

**Comment by Pacomatic1** [comment]
Just came to say that all the stuff in LBP2 also applies to LBP3; disabling MLAA improves performance and upscaling and all that.
Sadly, using the patch provided to disable MLAA on LBP3 also disables other effects.

**Comment by rleonever** [comment]
Hello, is this still an issue?

**Comment by Pacomatic1** [comment]
Yes.

**Comment by kd-11** [comment]
A decent number of these are fixed in https://github.com/RPCS3/rpcs3/pull/19013 when the 'blit engine upscaling' option is enabled.

**Comment by asxhun** [comment]
Upscaling on LBP2 (and possibly LBP3) is now fixed by #19013 while using the new GPU option "Allow Blit Engine Upscaling" since it's no longer required to disable MLAA. I'd expect #14055 to be fixed as well and also, I think forcing CPU Blit Emulation is no longer necessary on these games, but hopefully someone can correct me on that one.

**Comment by Pacomatic1** [comment]
I will test this once I get home.

**Comment by Pacomatic1** [comment]
...i can't find the option.
WIn11 @ latest.

**Comment by kd-11** [comment]
The option was removed. It is now enabled by default.

**Comment by Pacomatic1** [comment]
<img width="1919" height="1079" alt="Image" src="https://github.com/user-attachments/assets/9ccb2c4b-dec2-4eae-a502-4c758e81d7b5" />

It... kind of, works?
The 2D, like it used to, renders in native 1080p, but the 3D looks to render at its original resolution, and have itself upscaled either before or after the MLAA.
Better, but I don't think this was the goal.

I am on 1080p, but the same happens at 2160p as well.

**Comment by dfread** [comment]
Wanted to add Killzone 2 also has a similar issue; the post processing effects (motion blur, DOF effect on gun etc..) doesn't scale with Resolution Scale. Changing the Resolution Scale Threshold doesn't seem to work.
